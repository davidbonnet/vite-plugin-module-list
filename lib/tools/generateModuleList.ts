import { dirname, relative, resolve } from "path";

import { COMMENT } from "../constants.js";
import type { Mode } from "../types";

import { formatFilePathExtension } from "./formatFilePathExtension.js";
import { formatRelativePath } from "./formatRelativePath.js";
import { formatValue } from "./formatValue.js";
import { identifierFromFilePath } from "./identifierFromFilePath.js";

export function generateModuleList(
  filePathList: readonly string[],
  rootPath: string,
  outputPath: string,
  mode: Mode,
) {
  const outputRootPath = dirname(outputPath);
  switch (mode.language) {
    case "ts":
    case "js": {
      if (mode.dynamic) {
        const moduleList = `[\n${filePathList
          .map((filePath) => {
            const relativeFilePath = relative(
              outputRootPath,
              resolve(
                rootPath,
                formatFilePathExtension(filePath, mode.extension),
              ),
            );
            return `  { path: ${formatValue(
              relativeFilePath,
            )}, module: () => import(${formatRelativePath(
              relativeFilePath,
            )}) }`;
          })
          .join(",\n")}\n]`;
        return `// ${COMMENT}\nexport default ${moduleList};\n`;
      }
      const moduleList = filePathList
        .map((filePath) => {
          const relativeFilePath = relative(
            outputRootPath,
            resolve(
              rootPath,
              formatFilePathExtension(filePath, mode.extension),
            ),
          );
          return `export${
            mode.language === "ts" && mode.type ? " type" : ""
          } { ${identifierFromFilePath(filePath)} } from ${formatRelativePath(
            relativeFilePath,
          )}`;
        })
        .join("\n");
      return `// ${COMMENT}\n${moduleList}\n`;
    }
    case "css": {
      const moduleList = filePathList
        .map((filePath) => {
          const relativeFilePath = relative(
            outputRootPath,
            resolve(rootPath, filePath),
          );
          return `@import ${formatRelativePath(relativeFilePath)};`;
        })
        .join("\n");
      return `/* ${COMMENT} */\n${moduleList}\n`;
    }
    default:
      throw new Error(`Unsupported language '${(mode as any).language}'`);
  }
}
