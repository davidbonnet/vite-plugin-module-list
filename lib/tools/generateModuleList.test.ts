import { expect, test } from "vitest";

import { generateModuleList } from "./generateModuleList.js";

test("generates a dynamic module list for TypeScript", async () => {
  expect(
    generateModuleList(
      [
        "/root/path/lib/a.ts",
        "/root/path/lib/b.ts",
        "/root/path/lib/c.ts",
        "/root/path/lib/d/1.ts",
        "/root/path/lib/d/2.ts",
        "/root/path/lib/d/3.ts",
      ],
      "/root/path",
      "/root/path/lib.ts",
      { dynamic: true, extension: "js", language: "ts" },
    ),
  ).toMatchSnapshot();
});

test("generates a dynamic module list for JavaScript", async () => {
  expect(
    generateModuleList(
      [
        "/root/path/lib/a.ts",
        "/root/path/lib/b.ts",
        "/root/path/lib/c.ts",
        "/root/path/lib/d/1.ts",
        "/root/path/lib/d/2.ts",
        "/root/path/lib/d/3.ts",
      ],
      "/root/path",
      "/root/path/lib.ts",
      { dynamic: true, language: "js" },
    ),
  ).toMatchSnapshot();
});

test("generates a module list for TypeScript", async () => {
  expect(
    generateModuleList(
      [
        "/root/path/lib/a.ts",
        "/root/path/lib/b.ts",
        "/root/path/lib/c.ts",
        "/root/path/lib/d/1.ts",
        "/root/path/lib/d/2.ts",
        "/root/path/lib/d/3.ts",
      ],
      "/root/path",
      "/root/path/lib.ts",
      { extension: true, language: "ts" },
    ),
  ).toMatchSnapshot();
});

test("generates a module list for JavaScript", async () => {
  expect(
    generateModuleList(
      [
        "/root/path/lib/a.js",
        "/root/path/lib/b.js",
        "/root/path/lib/c.js",
        "/root/path/lib/d/1.js",
        "/root/path/lib/d/2.js",
        "/root/path/lib/d/3.js",
      ],
      "/root/path",
      "/root/path/lib.js",
      { extension: true, language: "js" },
    ),
  ).toMatchSnapshot();
});

test("generates a types module list for TypeScript", async () => {
  expect(
    generateModuleList(
      [
        "/root/path/lib/a.ts",
        "/root/path/lib/b.ts",
        "/root/path/lib/c.ts",
        "/root/path/lib/d/1.ts",
        "/root/path/lib/d/2.ts",
        "/root/path/lib/d/3.ts",
      ],
      "/root/path",
      "/root/path/lib.ts",
      { extension: true, language: "ts", type: true },
    ),
  ).toMatchSnapshot();
});
