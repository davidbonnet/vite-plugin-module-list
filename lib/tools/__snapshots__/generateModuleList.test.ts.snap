// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`generates a dynamic module list for JavaScript 1`] = `
"// File automatically generated by \`vite-plugin-module-list\`
export default [
  { path: "lib/a", module: () => import("./lib/a") },
  { path: "lib/b", module: () => import("./lib/b") },
  { path: "lib/c", module: () => import("./lib/c") },
  { path: "lib/d/1", module: () => import("./lib/d/1") },
  { path: "lib/d/2", module: () => import("./lib/d/2") },
  { path: "lib/d/3", module: () => import("./lib/d/3") }
];
"
`;

exports[`generates a dynamic module list for TypeScript 1`] = `
"// File automatically generated by \`vite-plugin-module-list\`
export default [
  { path: "lib/a.js", module: () => import("./lib/a.js") },
  { path: "lib/b.js", module: () => import("./lib/b.js") },
  { path: "lib/c.js", module: () => import("./lib/c.js") },
  { path: "lib/d/1.js", module: () => import("./lib/d/1.js") },
  { path: "lib/d/2.js", module: () => import("./lib/d/2.js") },
  { path: "lib/d/3.js", module: () => import("./lib/d/3.js") }
];
"
`;

exports[`generates a module list for JavaScript 1`] = `
"// File automatically generated by \`vite-plugin-module-list\`
export { a } from "./lib/a.js"
export { b } from "./lib/b.js"
export { c } from "./lib/c.js"
export { 1 } from "./lib/d/1.js"
export { 2 } from "./lib/d/2.js"
export { 3 } from "./lib/d/3.js"
"
`;

exports[`generates a module list for TypeScript 1`] = `
"// File automatically generated by \`vite-plugin-module-list\`
export { a } from "./lib/a.ts"
export { b } from "./lib/b.ts"
export { c } from "./lib/c.ts"
export { 1 } from "./lib/d/1.ts"
export { 2 } from "./lib/d/2.ts"
export { 3 } from "./lib/d/3.ts"
"
`;

exports[`generates a types module list for TypeScript 1`] = `
"// File automatically generated by \`vite-plugin-module-list\`
export type { a } from "./lib/a.ts"
export type { b } from "./lib/b.ts"
export type { c } from "./lib/c.ts"
export type { 1 } from "./lib/d/1.ts"
export type { 2 } from "./lib/d/2.ts"
export type { 3 } from "./lib/d/3.ts"
"
`;
